clippy_lints/src/use_self.rs

```diff
@@ -192,18 +198,9 @@ impl<'tcx> LateLintPass<'tcx> for UseSelf {
         }
     }
 
-    fn check_body(&mut self, _: &LateContext<'_>, _: &hir::Body<'_>) {
-        // `lower_ty` cannot be called in `Body`s or it will panic (sometimes). But in bodies
-        // we can use `cx.typeck_results.node_type(..)` to get the `ty::Ty` from a `hir::Ty`.
-        // However the `node_type()` method can *only* be called in bodies.
-        if let Some(&mut StackItem::Check { ref mut in_body, .. }) = self.stack.last_mut() {
-            *in_body = in_body.saturating_add(1);
-        }
-    }
-
-    fn check_body_post(&mut self, _: &LateContext<'_>, _: &hir::Body<'_>) {
-        if let Some(&mut StackItem::Check { ref mut in_body, .. }) = self.stack.last_mut() {
-            *in_body = in_body.saturating_sub(1);
+    fn check_impl_item_post(&mut self, _: &LateContext<'_>, _: &hir::ImplItem<'_>) {
+        if let Some(StackItem::NoCheck) = self.stack.last() {
+            self.stack.pop();
         }
     }
```

Comment:
Sorry for the delay!

We'll need to also check `if impl_item.span.from_expansion()` here so that we only pop entries we added in `check_impl_item`

Without it we would run into

```rust
struct A;

impl A {
    fn a() {
        struct B;

        impl Iterator for &B { // pushes a NoCheck
            type Item = A; // Pops the NoCheck
        
            fn next(&mut self) -> Option<A> { // Lints A -> Self
                Some(A)
            }
        }
    }
}
```

Could you add a test for this too

Reply:
Anywhere in `tests/ui/use_self.rs` would be fine

Reply:
Done.